src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs

```diff
@@ -26,21 +27,34 @@ public void Initialize(JobHost host)
         {
         }
 
-        public Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, TimeSpan timeoutGracePeriod)
+        public async Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, TimeSpan timeoutGracePeriod)
         {
+            FunctionTimeoutException timeoutException = exceptionInfo.SourceException as FunctionTimeoutException;
+
+            if (timeoutException?.FunctionTask != null)
+            {
+                // We may double the timeoutGracePeriod here by first waiting to see if the iniital
+                // function task that started the exception has completed.
+                Task completedTask = await Task.WhenAny(timeoutException.FunctionTask, Task.Delay(timeoutGracePeriod));
+
+                // If the function task has completed, simply return. We've already logged the timeout.
+                if (completedTask == timeoutException.FunctionTask)
+                {
+                    return;
+                }
+            }
+
             LogErrorAndFlush("A function timeout has occurred. Host is shutting down.", exceptionInfo.SourceException);
 
             // We can't wait on this as it may cause a deadlock if the timeout was fired
             // by a Listener that cannot stop until it has completed.
-            Task taskIgnore = _manager.StopAsync();
+            Task taskIgnoreStop = _manager.StopAsync();
 
-            // Give the manager some time to shut down gracefully.
-            Task.Delay(timeoutGracePeriod).ContinueWith((ct) =>
+            // Give the manager and all running tasks some time to shut down gracefully.
+            Task taskIgnoreShutdown = Task.Delay(timeoutGracePeriod).ContinueWith((ct) =>
```

Comment:
We should be able to just `await` on the `Task.Delay` here instead of using a continuation, right?